src/ucc/api/ucc.h

```diff
@@ -1012,8 +1012,42 @@ typedef struct ucc_team_p2p_conn {
  *  @ingroup UCC_TEAM_DT
  */
 typedef enum {
-    UCC_COLLECTIVE_POST_ORDERED     = 0,
-    UCC_COLLECTIVE_POST_UNORDERED   = 1
+
+    /**
+      * When set to this value, the collective participants shall invoke the operation
+      * in the same order.
+      */
+    UCC_COLLECTIVE_POST_ORDERED             = 0,
+
+    /**
+      * When set to this value, the collective participants shall invoke the operation
+      * in any order.
+      */
+    UCC_COLLECTIVE_POST_UNORDERED           = 1,
+
+    /**
+      * When set to this value, the collective participants shall initialize the operation
+      * in the same order.
+      */
+    UCC_COLLECTIVE_INIT_ORDERED             = 2,
+
+    /**
+      * When set to this value, the collective participants shall initialize the operation
+      * in any order.
+      */
+    UCC_COLLECTIVE_INIT_UNORDERED           = 3,
+
+    /**
+      * When set to this value, the collective participants shall initialize and
+      * invoke the operation in the same order.
+      */
+    UCC_COLLECTIVE_INIT_AND_POST_ORDERED    = 4,
+
+    /**
+      * When set to this value, the collective participants shall initialize and
+      * invoke the operation in any order.
+      */
+    UCC_COLLECTIVE_INIT_AND_POST_UNORDERED  = 5
 } ucc_post_ordering_t;
```

Comment:
is there case where INIT is ordered but post is unordered?

Reply:
is it make sense to use bitmask instead of enum here?

Reply:
> is there case where INIT is ordered but post is unordered?

The MPI persistent collectives in MPI 4.0 require those semantics. When we set to the value "UCC_COLLECTIVE_POST_ORDERED", there is no constraint on the ordering of Init. Should we specify the order for init in this case? I left it undefined intentionally.

Reply:
> is it make sense to use bitmask instead of enum here?

I don't have a preference.

Reply:
> is there case where INIT is ordered but post is unordered?

This semantic is required for MPI 4.0's persistence collectives. When UCC_COLLECTIVE_INIT_ORDERED is set, the post can be ordered or unordered. Is this enough or we should tighten it up?

Reply:
@manjugv one thing i don't understand: if, say, user sets "UCC_COLLECTIVE_INIT_ORDERED" only. Do this imply POST = unordered (basically post can by "any" but for implementation this means it has to support potentially unordered posts) ?

Reply:
Correct, it can be either ordered or unordered. Ordered Init and unordered post is needed for MPI's persistent collective. The question I have, do we need to be explicit about the post being ordered or unordered?

Reply:
@vspetrov Did we agree on leaving the semantics as it is or should we add INIT_AND_UNORDERED?

Reply:
UCC WG: Agreed not to add.